hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/spark/sql/hudi/command/TruncateHoodieTableCommand.scala

```diff
@@ -18,115 +18,69 @@
+    if (table.partitionColumnNames.isEmpty && specs.isDefined) {
+      throw new AnalysisException(
+        s"Operation not allowed: TRUNCATE TABLE ... PARTITION is not supported " +
+          s"for tables that are not partitioned: $tableIdentWithDB")
+    }
 
-      // Also try to drop the contents of the table from the columnar cache
-      try {
-        spark.sharedState.cacheManager.uncacheQuery(spark.table(table.identifier), cascade = true)
-      } catch {
-        case NonFatal(_) =>
-      }
+    val basePath = hoodieCatalogTable.tableLocation
+    val properties = hoodieCatalogTable.tableConfig.getProps
+    val hadoopConf = sparkSession.sessionState.newHadoopConf()
 
-      if (table.stats.nonEmpty) {
-        // empty table after truncation
-        val newStats = CatalogStatistics(sizeInBytes = 0, rowCount = Some(0))
-        catalog.alterTableStats(tableIdentifier, Some(newStats))
-      }
-      Seq.empty[Row]
-    } catch {
-      // TruncateTableCommand will delete the related directories first, and then refresh the table.
-      // It will fail when refresh table, because the hudi meta directory(.hoodie) has been deleted at the first step.
-      // So here ignore this failure, and refresh table later.
-      case NonFatal(e) =>
-        throw new AnalysisException(s"Exception when attempting to truncate table ${tableIdentifier.quotedString}: " + e)
-    }
+    if (specs.isEmpty) {
+      val targetPath = new Path(basePath)
+      val engineContext = new HoodieSparkEngineContext(sparkSession.sparkContext)
+      val fs = FSUtils.getFs(basePath, sparkSession.sparkContext.hadoopConfiguration)
+      FSUtils.deleteDir(engineContext, fs, targetPath, sparkSession.sparkContext.defaultParallelism)
 
-    // If we have not specified the partition, truncate will delete all the data in the table path
-    // include the hoodie.properties. In this case we should reInit the table.
-    if (partitionSpec.isEmpty) {
-      val hadoopConf = spark.sessionState.newHadoopConf()
       // ReInit hoodie.properties
       HoodieTableMetaClient.withPropertyBuilder()
         .fromProperties(properties)
         .initTable(hadoopConf, hoodieCatalogTable.tableLocation)
-    }
 
-    // After deleting the data, refresh the table to make sure we don't keep around a stale
-    // file relation in the metastore cache and cached table data in the cache manager.
-    spark.catalog.refreshTable(hoodieCatalogTable.table.identifier.quotedString)
-    Seq.empty[Row]
+      // After deleting the data, refresh the table to make sure we don't keep around a stale
+      // file relation in the metastore cache and cached table data in the cache manager.
+      sparkSession.catalog.refreshTable(table.identifier.quotedString)
+      Seq.empty[Row]
+    } else {
+      AlterHoodieTableDropPartitionCommand(tableIdentifier,
+        specs = Seq(specs.get),
+        ifExists = false,
+        purge = false,
+        retainData = false)
+        .run(sparkSession)
```

Comment:
I suggest not to mix Drop with Truncate command implementation. e.g. when you do logging, it's may log from Drop but users are actually running Truncate, which is misleading. Also isolating the implementation makes the logic easier to understand.